# **Mielage Prediction**

-------------

## **Objective**

To predict the mielage of of a vehicle from the given data-set.

## **Data Source**

https://github.com/YBI-Foundation/Dataset/raw/main/MPG.csv

## **Import Library**

In [100]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_absolute_percentage_error

## **Import Data**

In [101]:
mielage = pd.read_csv('https://github.com/YBI-Foundation/Dataset/raw/main/MPG.csv')

## **Describe Data**

In [102]:
mielage.describe()

mpg   cylinders  displacement  horsepower       weight  \
count  398.000000  398.000000    398.000000  392.000000   398.000000   
mean    23.514573    5.454774    193.425879  104.469388  2970.424623   
std      7.815984    1.701004    104.269838   38.491160   846.841774   
min      9.000000    3.000000     68.000000   46.000000  1613.000000   
25%     17.500000    4.000000    104.250000   75.000000  2223.750000   
50%     23.000000    4.000000    148.500000   93.500000  2803.500000   
75%     29.000000    8.000000    262.000000  126.000000  3608.000000   
max     46.600000    8.000000    455.000000  230.000000  5140.000000   

       acceleration  model_year  
count    398.000000  398.000000  
mean      15.568090   76.010050  
std        2.757689    3.697627  
min        8.000000   70.000000  
25%       13.825000   73.000000  
50%       15.500000   76.000000  
75%       17.175000   79.000000  
max       24.800000   82.000000

## **Data Visualization**

In [103]:
mielage.head()

mpg  cylinders  displacement  horsepower  weight  acceleration  \
0  18.0          8         307.0       130.0    3504          12.0   
1  15.0          8         350.0       165.0    3693          11.5   
2  18.0          8         318.0       150.0    3436          11.0   
3  16.0          8         304.0       150.0    3433          12.0   
4  17.0          8         302.0       140.0    3449          10.5   

   model_year origin                       name  
0          70    usa  chevrolet chevelle malibu  
1          70    usa          buick skylark 320  
2          70    usa         plymouth satellite  
3          70    usa              amc rebel sst  
4          70    usa                ford torino

In [104]:
mielage.tail()

mpg  cylinders  displacement  horsepower  weight  acceleration  \
393  27.0          4         140.0        86.0    2790          15.6   
394  44.0          4          97.0        52.0    2130          24.6   
395  32.0          4         135.0        84.0    2295          11.6   
396  28.0          4         120.0        79.0    2625          18.6   
397  31.0          4         119.0        82.0    2720          19.4   

     model_year  origin             name  
393          82     usa  ford mustang gl  
394          82  europe        vw pickup  
395          82     usa    dodge rampage  
396          82     usa      ford ranger  
397          82     usa       chevy s-10

## **Data Preprocessing**

In [105]:
mielage = mielage.dropna()

In [106]:
mielage.info()

<class 'pandas.core.frame.DataFrame'>
Index: 392 entries, 0 to 397
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   mpg           392 non-null    float64
 1   cylinders     392 non-null    int64  
 2   displacement  392 non-null    float64
 3   horsepower    392 non-null    float64
 4   weight        392 non-null    int64  
 5   acceleration  392 non-null    float64
 6   model_year    392 non-null    int64  
 7   origin        392 non-null    object 
 8   name          392 non-null    object 
dtypes: float64(4), int64(3), object(2)
memory usage: 30.6+ KB


In [107]:
mielage.columns

Index(['mpg', 'cylinders', 'displacement', 'horsepower', 'weight',
       'acceleration', 'model_year', 'origin', 'name'],
      dtype='object')

In [108]:
mielage['origin'].value_counts()

origin
usa       245
japan      79
europe     68
Name: count, dtype: int64

In [109]:
mielage.replace({'origin':{'usa':0, 'japan':1, 'europe':2}}, inplace=True)

C:\Users\arvin\AppData\Local\Temp\ipykernel_31852\1762271917.py:1: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  mielage.replace({'origin':{'usa':0, 'japan':1, 'europe':2}}, inplace=True)


In [110]:
mielage.head()

mpg  cylinders  displacement  horsepower  weight  acceleration  \
0  18.0          8         307.0       130.0    3504          12.0   
1  15.0          8         350.0       165.0    3693          11.5   
2  18.0          8         318.0       150.0    3436          11.0   
3  16.0          8         304.0       150.0    3433          12.0   
4  17.0          8         302.0       140.0    3449          10.5   

   model_year  origin                       name  
0          70       0  chevrolet chevelle malibu  
1          70       0          buick skylark 320  
2          70       0         plymouth satellite  
3          70       0              amc rebel sst  
4          70       0                ford torino

## **Define Target Variable (y) and Feature Variables (X)**

In [111]:
y = mielage['mpg']
X = mielage[['cylinders', 'displacement', 'horsepower', 'weight','acceleration', 'model_year', 'origin']]

## **Train Test Split**

In [112]:
X_train, X_test, y_train, y_test = train_test_split(X, y)
X_train.shape, X_test.shape, y_train.shape, y_test.shape

((294, 7), (98, 7), (294,), (98,))

## **Modeling**

In [113]:
model = LinearRegression()
model.fit(X_train, y_train)

LinearRegression()

## **Model Evaluation**

In [114]:
model.intercept_

-18.065949384481346

In [115]:
model.coef_

array([-0.40547381,  0.01282148,  0.0057375 , -0.00668133,  0.07839666,
        0.77157432,  1.30026162])

## **Prediction**

In [116]:
y_pred = model.predict(X_test)
print(y_pred)

[30.16037424 22.08091622 27.09243031 26.8813504  30.88636765 24.34949193
 25.26005903 28.97069038 29.65387728 33.3365566  25.97110688 19.49032008
 33.61095175 31.54091847 15.96799932 23.69978336 15.29216688 23.98278871
 21.57406387 19.75453168 27.86321034 17.23889389 27.68002031 17.90038703
 32.9497601  31.3521037  28.72433835 21.90399121 23.41658851 11.89746504
 26.20449908 29.17138294 13.31786039 26.80215222 34.37346019 12.15434257
 10.93549766 33.08815119 19.50542279 30.83697649 28.78300689 32.11965104
 32.32506341 25.11183638 32.04615558 20.6456636  28.01914551 28.04781474
 19.85207645 20.69380179 21.6258033  30.56258573 31.23265859 28.06739433
 20.76162605 24.34406653 16.00832489 22.13039516 25.08850572 28.81351041
 28.41509197 18.89394557 28.13845665 25.94872195 23.46475731 12.29058339
 30.47862385 12.7810166  28.20957974 24.1645254  32.94790602 30.49213682
 32.0458333  20.08649236 30.91449912 17.07826785 16.97817757  9.92301709
 25.35016281 25.93183591 27.67631327 32.99318411 28

In [117]:
accuracy = 100 - mean_absolute_percentage_error(y_pred, y_test)
print(accuracy)

99.89326699619946


## **Explaination**

With this the mielage was predicted successfully to an accuracy of 99.89%